### ISSUE TITLE
UI for lots of repos

### ISSUE THREAD

**Issue by ewjoachim** [git-cola/git-cola#1090]
Hey there,
I'm a long time fan of the project.

In the last months, I've became involved in many different repositories (30+). I used to keep all the repos I interacted with in "Favorites" but it's becoming less and less usable. I can always open a repo with ctrl+o but that's not the most practical thing to do. I'm wondering if there's something I might contribute (or some existing feature I missed) that could help for my case. Ideally I'd use a shortcut, type a few letters of the repo I'm interested in, and then either click or use the arrow keys to select it. A "repo quick-switcher", kind of like what there is in recent editors (usually mapped to `cmd/ctrl+p`).

So is there something of the likes already existing ? If not, are there some places in the codebase you recommand as inspiration if I were to build it ?

**Comment by davvid** [comment]
This seems like it'd be pretty useful. The "Recent" and "Favorites" widgets do let you type in stuff and it'll select one of the items in the list. Then, hitting "enter" will switch to that repo. It'd be cool if this were available globally through a shortcut like you suggested.

**Comment by ewjoachim** [comment]
Hey :) Thanks for you answer. I'm not sure I will have time to tackle this now. If anyone is interested, please feel free!

**Comment by nakanoi** [comment]
Hi @davvid @ewjoachim !
I tried to implement these features in this PR #1282.
- filtering repositories on bookmarks(recent) widget
- global shortcut for filtering repositroies

I would be very happy if you could take a look at it.
Thank you so much!

**Comment by ewjoachim** [comment]
[Screencast from 13-01-2023 13:28:39.webm](https://user-images.githubusercontent.com/1457576/212320577-162245e1-6daf-4e85-a39e-bcd3d95745ca.webm)

Nice !
2 very optional comments:
- The way "ctrl+p" menus usually work is that when you press enter, the first element is activated. In this case, you need to press down. Ideally, make the first element already selected
- It's filtering by trying to search the string you typed inside each repo. It would be marginally better if it was looking for any repo that have the same letters as you typed, but with potentially other letters in between. For example, if we type `pycocoac`, `python-coverage-comment-action` should be available (if we want it to work the same way as the ctrl+p or ctrl+t menus from Sublime text, VS Code, Slack, Discord etc)

**Comment by ewjoachim** [comment]
For the 2nd item, I think it would be simple as
```python
self.proxy_model.setFilterRegExp(".*".join(input_text))
```
for this line: https://github.com/git-cola/git-cola/pull/1282/files#diff-fd2f970770c2816e8cb76d8780ddf3332353ead64c3c563ff1b2c74fed11b032R85

**Comment by ewjoachim** [comment]
Yep, I've tested, that works.

**Comment by nakanoi** [comment]
@ewjoachim 
Thank you for reviewing!

> For the 2nd item, I think it would be simple as
> 
> ```python
> self.proxy_model.setFilterRegExp(".*".join(input_text))
> ```

Wow, I couldn't come up with such a briliant idea! Thank you for suggestions.
I'll improve the code with your suggestions!

**Comment by nakanoi** [comment]
I implemented the suggestions!  - initial defeault selection & potentially character regex.

---
you're probably aware of this, but I also implemented following comment.
> The "Recent" and "Favorites" widgets do let you type in stuff and it'll select one of the items in the list. Then, hitting "enter" will switch to that repo.

Focus on bookmarks/recent widget, type in some characters, and filter items with them. Then press enter switches current repository as following video.

[!(filtering webm)](https://user-images.githubusercontent.com/72122101/212522325-fe5be429-283d-42a0-8d21-4baaf8978d81.webm)
